File: hydra-haskell/src/main/haskell/Hydra/Reduction.hs
-- | Functions for reducing terms and types, i.e. performing computations

module Hydra.Reduction where

import Hydra.Basics
import Hydra.Strip
import Hydra.Compute
import Hydra.Core
import Hydra.CoreDecoding
import Hydra.Extras
import Hydra.Graph
import Hydra.Kv
import Hydra.Lexical
import Hydra.Rewriting
import Hydra.Tier1
import Hydra.Tier2
import qualified Hydra.Dsl.Expect as Expect
import qualified Hydra.Dsl.Terms as Terms

import qualified Control.Monad as CM
import qualified Data.List as L
import qualified Data.Map as M
import qualified Data.Set as S
import qualified Data.Maybe as Y


alphaConvert :: Name -> Term -> Term -> Term
alphaConvert vold tnew = rewriteTerm rewrite id
  where
    rewrite recurse term = case term of
      TermFunction (FunctionLambda (Lambda v body)) -> if v == vold
        then term
        else recurse term
      TermVariable v -> if v == vold then tnew else TermVariable v
      _ -> recurse term

-- For demo purposes. This should be generalized to enable additional side effects of interest.
countPrimitiveInvocations :: Bool
countPrimitiveInvocations = True

-- A term evaluation function which is alternatively lazy or eager
reduceTerm :: Bool -> M.Map Name (Term) -> Term -> Flow (Graph) (Term)
reduceTerm eager env = rewriteTermM mapping pure
  where
    reduce eager = reduceTerm eager M.empty

    mapping recurse mid = do
      inner <- if doRecurse eager mid then recurse mid else pure mid
      applyIfNullary eager inner []

    doRecurse eager term = eager && case term of
      TermFunction (FunctionLambda _) -> False
      _ -> True

    -- Reduce an argument only if evaluation is lazy (i.e. the argument may not already have been reduced)
    reduceArg eager arg = if eager then pure arg else reduce False arg

    applyToArguments fun args = case args of
      [] -> fun
      (h:r) -> applyToArguments (Terms.apply fun h) r

    replaceFreeName toReplace replacement = rewriteTerm mapping id
      where
        mapping recurse inner = case inner of
          TermFunction (FunctionLambda (Lambda param body)) -> if param == toReplace then inner else recurse inner
          TermVariable name -> if name == toReplace then replacement else inner
          _ -> recurse inner

    applyIfNullary eager original args = case stripTerm original of
      TermApplication (Application fun arg) -> applyIfNullary eager fun (arg:args)
      TermFunction fun -> case fun of
        FunctionElimination elm -> case args of
          [] -> pure original
          -- Reduce the argument prior to application, regardless of laziness
          (arg:remainingArgs) -> do
            reducedArg <- reduceArg eager $ stripTerm arg
            reducedResult <- applyElimination elm reducedArg >>= reduce eager
            applyIfNullary eager reducedResult remainingArgs
        FunctionLambda (Lambda param body) -> case args of
          [] -> pure original
          (arg:remainingArgs) -> do
            reducedArg <- reduce eager $ stripTerm arg
            reducedResult <- reduce eager $ replaceFreeName param reducedArg body
            applyIfNullary eager reducedResult remainingArgs
        FunctionPrimitive name -> do
          prim <- requirePrimitive name
          let arity = primitiveArity prim
          if arity > L.length args
            -- Not enough arguments available; back out
            then return $ applyToArguments original args
            else do
              let argList = L.take arity args
              let remainingArgs = L.drop arity args
              reducedArgs <- CM.mapM (reduceArg eager) argList
              reducedResult <- primitiveImplementation prim reducedArgs >>= reduce eager
              applyIfNullary eager reducedResult remainingArgs
      TermVariable _ -> pure $ applyToArguments original args -- TODO: dereference variables
      _ -> pure $ applyToArguments original args

    applyElimination elm reducedArg = case elm of
      EliminationList _ -> fail "list eliminations are unsupported"
      EliminationOptional _ -> fail "optional eliminations are unsupported"
      EliminationRecord proj -> do
        fields <- Expect.recordWithName (projectionTypeName proj) $ stripTerm reducedArg
        let matchingFields = L.filter (\f -> fieldName f == projectionField proj) fields
        if L.null matchingFields
          then fail $ "no such field: " ++ unName (projectionField proj) ++ " in " ++ unName (projectionTypeName proj) ++ " record"
          else pure $ fieldTerm $ L.head matchingFields
      EliminationUnion (CaseStatement name def fields) -> do
        field <- Expect.injectionWithName name reducedArg
        let matchingFields = L.filter (\f -> fieldName f == fieldName field) fields
        if L.null matchingFields
          then case def of
            Just d -> pure d
            Nothing -> fail $ "no such field " ++ unName (fieldName field) ++ " in " ++ unName name ++ " case statement"
          else pure $ Terms.apply (fieldTerm $ L.head matchingFields) (fieldTerm field)
      EliminationWrap name -> Expect.wrap name reducedArg

-- Note: this is eager beta reduction, in that we always descend into subtypes,
--       and always reduce the right-hand side of an application prior to substitution
betaReduceType :: Type -> Flow (Graph) (Type)
betaReduceType typ = do
    g <- getState :: Flow (Graph) (Graph)
    rewriteTypeM mapExpr (pure . id) typ
  where
    mapExpr recurse t = do
        r <- recurse t
        case r of
          TypeApplication a -> reduceApp a
          t' -> pure t'
      where
        reduceApp (ApplicationType lhs rhs) = case lhs of
          TypeAnnotated (AnnotatedType t' ann) -> do
            a <- reduceApp $ ApplicationType t' rhs
            return $ TypeAnnotated $ AnnotatedType a ann
          TypeLambda (LambdaType v body) -> betaReduceType $ replaceFreeName v rhs body
          -- nominal types are transparent
          TypeVariable name -> do
            t' <- requireType name
            betaReduceType $ TypeApplication $ ApplicationType t' rhs

-- | Apply the special rules:
--     ((\x.e1) e2) == e1, where x does not appear free in e1
--   and
--     ((\x.e1) e2) = e1[x/e2]
--  These are both limited forms of beta reduction which help to "clean up" a term without fully evaluating it.
contractTerm :: Term -> Term
contractTerm = rewriteTerm rewrite id
  where
    rewrite recurse term = case rec of
        TermApplication (Application lhs rhs) -> case stripTerm lhs of
          TermFunction (FunctionLambda (Lambda v body)) -> if isFreeIn v body
            then body
            else alphaConvert v rhs body
          _ -> rec
        _ -> rec
      where
        rec = recurse term

-- Note: unused / untested
etaReduceTerm :: Term -> Term
etaReduceTerm term = case term of
    TermAnnotated (AnnotatedTerm term1 ann) -> TermAnnotated (AnnotatedTerm (etaReduceTerm term1) ann)
    TermFunction (FunctionLambda l) -> reduceLambda l
    _ -> noChange
  where
    reduceLambda (Lambda v body) = case etaReduceTerm body of
      TermAnnotated (AnnotatedTerm body1 ann) -> reduceLambda (Lambda v body1)
      TermApplication a -> reduceApplication a
        where
          reduceApplication (Application lhs rhs) = case etaReduceTerm rhs of
            TermAnnotated (AnnotatedTerm rhs1 ann) -> reduceApplication (Application lhs rhs1)
            TermVariable v1 -> if v == v1 && isFreeIn v lhs
              then etaReduceTerm lhs
              else noChange
            _ -> noChange
      _ -> noChange
    noChange = term

-- | Whether a term is closed, i.e. represents a complete program
termIsClosed :: Term -> Bool
termIsClosed = S.null . freeVariablesInTerm

-- | Whether a term has been fully reduced to a "value"
termIsValue :: Graph -> Term -> Bool
termIsValue g term = case stripTerm term of
    TermApplication _ -> False
    TermLiteral _ -> True
    TermFunction f -> functionIsValue f
    TermList els -> forList els
    TermMap map -> L.foldl
      (\b (k, v) -> b && termIsValue g k && termIsValue g v)
      True $ M.toList map
    TermOptional m -> case m of
      Nothing -> True
      Just term -> termIsValue g term
    TermRecord (Record _ fields) -> checkFields fields
    TermSet els -> forList $ S.toList els
    TermUnion (Injection _ field) -> checkField field
    TermVariable _ -> False
  where
    forList els = L.foldl (\b t -> b && termIsValue g t) True els
    checkField = termIsValue g . fieldTerm
    checkFields = L.foldl (\b f -> b && checkField f) True

    functionIsValue f = case f of
      FunctionElimination e -> case e of
        EliminationWrap _ -> True
        EliminationOptional (OptionalCases nothing just) -> termIsValue g nothing
          && termIsValue g just
        EliminationRecord _ -> True
        EliminationUnion (CaseStatement _ def cases) -> checkFields cases && (Y.maybe True (termIsValue g) def)
      FunctionLambda (Lambda _ body) -> termIsValue g body
      FunctionPrimitive _ -> True

File: hydra-java/src/main/java/hydra/Reduction.java
package hydra;

import hydra.compute.Flow;
import hydra.core.Application;
import hydra.core.CaseStatement;
import hydra.core.Elimination;
import hydra.core.Field;
import hydra.core.Injection;
import hydra.core.Lambda;
import hydra.core.Name;
import hydra.core.Projection;
import hydra.core.Record;
import hydra.core.Term;
import hydra.dsl.Flows;
import hydra.dsl.Terms;
import hydra.graph.Graph;
import hydra.tools.LList;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import static hydra.dsl.Flows.bind;
import static hydra.dsl.Flows.fail;
import static hydra.dsl.Flows.mapM;
import static hydra.dsl.Flows.pure;
import static hydra.dsl.Flows.unexpected;
import static hydra.extras.Extras.primitiveArity;
import static hydra.strip.Strip.stripTerm;


public class Reduction {
    private Reduction() {
    }

    private static  Flow<Graph, Term> applyElimination(Elimination elm, Term reducedArg) {
        return elm.accept(new Elimination.Visitor<Flow<Graph, Term>>() {
            @Override
            public Flow<Graph, Term> visit(Elimination.List instance) {
                throw new UnsupportedOperationException();
            }

            @Override
            public Flow<Graph, Term> visit(Elimination.Optional instance) {
                throw new UnsupportedOperationException();
            }

            @Override
            public Flow<Graph, Term> visit(Elimination.Product instance) {
                throw new UnsupportedOperationException();
            }

            @Override
            public Flow<Graph, Term> visit(Elimination.Record elim) {
                return reducedArg.accept(new Term.PartialVisitor<Flow<Graph, Term>>() {
                    @Override
                    public Flow<Graph, Term> otherwise(Term instance) {
                        return unexpected("record", instance);
                    }

                    @Override
                    public Flow<Graph, Term> visit(Term.Record instance) {
                        Record record = instance.value;
                        Projection proj = elim.value;
                        if (record.typeName.equals(proj.typeName)) {
                            for (Field field : record.fields) {
                                if (field.name.equals(proj.field)) {
                                    return pure(field.term);
                                }
                            }
                            return fail("no such field: " + proj.field + " in " + record.typeName + " record");
                        } else {
                            return fail("tried to project a " + proj.typeName + " field out of a " + record.typeName
                                    + " record");
                        }
                    }
                });
            }

            @Override
            public Flow<Graph, Term> visit(Elimination.Union elim) {
                return reducedArg.accept(new Term.PartialVisitor<Flow<Graph, Term>>() {
                    @Override
                    public Flow<Graph, Term> otherwise(Term instance) {
                        return unexpected("injection", instance);
                    }

                    @Override
                    public Flow<Graph, Term> visit(Term.Union instance) {
                        CaseStatement cases = elim.value;
                        Injection inj = instance.value;
                        if (cases.typeName.equals(inj.typeName)) {
                            for (Field field : cases.cases) {
                                if (field.name.equals(inj.field.name)) {
                                    return pure(Terms.apply(field.term, inj.field.term));
                                }
                            }
                            return cases.default_.isPresent() ? pure(cases.default_.get())
                                    : fail("no such field " + inj.field.name
                                    + " in " + cases.typeName + " case statement");
                        } else {
                            return fail("tried to match a " + inj.typeName + " injection as " + cases.typeName);
                        }
                    }
                });
            }

            @Override
            public Flow<Graph, Term> visit(Elimination.Wrap instance) {
                return reducedArg.accept(new Term.PartialVisitor<Flow<Graph, Term>>() {
                    @Override
                    public Flow<Graph, Term> otherwise(Term instance) {
                        return unexpected("wrapped term", instance);
                    }

                    @Override
                    public Flow<Graph, Term> visit(Term.Wrap wrapped) {
                        Name fname = instance.value;
                        Name aname = wrapped.value.typeName;
                        return fname.equals(aname) ? pure(wrapped.value.object)
                                : fail("tried to unwrap an instance of " + aname + " as an instance of " + fname);
                    }
                });
            }
        });
    }

    private static  Flow<Graph, Term> applyIfNullary(boolean eager, Term original, LList<Term> args) {
        return stripTerm(original).accept(new Term.PartialVisitor<Flow<Graph, Term>>() {
            @Override
            public Flow<Graph, Term> otherwise(Term instance) {
                return pure(applyToArguments(original, args));
            }

            @Override
            public Flow<Graph, Term> visit(Term.Application instance) {
                Application app = instance.value;
                return applyIfNullary(eager, app.function, LList.push(app.argument, args));
            }

            @Override
            public Flow<Graph, Term> visit(Term.Function instance) {
                hydra.core.Function fun = instance.value;
                return fun.accept(new hydra.core.Function.Visitor<Flow<Graph, Term>>() {
                    @Override
                    public Flow<Graph, Term> visit(hydra.core.Function.Elimination instance) {
                        if (LList.isEmpty(args)) {
                            return pure(original);
                        } else {
                            Elimination elm = instance.value;
                            Term arg = args.first;
                            LList<Term> remainingArgs = args.rest;
                            // Reduce the argument prior to application, regardless of laziness
                            return bind(reduceArg(eager, stripTerm(arg)),
                                    reducedArg -> bind(applyElimination(elm, reducedArg),
                                            result -> bind(reduce(eager, result),
                                                    reducedResult -> applyIfNullary(eager, reducedResult,
                                                            remainingArgs))));
                        }
                    }

                    @Override
                    public Flow<Graph, Term> visit(hydra.core.Function.Lambda instance) {
                        if (LList.isEmpty(args)) {
                            return pure(original);
                        } else {
                            Lambda lam = instance.value;
                            Term arg = args.first;
                            LList<Term> remainingArgs = args.rest;
                            return bind(reduce(eager, stripTerm(arg)),
                                    reducedArg -> {
                                        Term result = replaceFreeName(lam.parameter, reducedArg, lam.body);
                                        return bind(reduce(eager, result),
                                                reducedResult -> applyIfNullary(eager, reducedResult, remainingArgs));
                                    });
                        }
                    }

                    @Override
                    public Flow<Graph, Term> visit(hydra.core.Function.Primitive instance) {
                        return bind(Lexical.requirePrimitive(instance.value), prim -> {
                            int arity = primitiveArity(prim);
                            if (arity <= LList.length(args)) {
                                List<Term> argList = LList.take(arity, args);
                                Flow<Graph, List<Term>> reducedArgs = mapM(argList, a -> reduceArg(eager, a));
                                LList<Term> remainingArgs = LList.drop(arity, args);
                                return bind(reducedArgs, rargs -> bind(prim.implementation.apply(rargs),
                                        result -> bind(reduce(eager, result),
                                                reducedResult -> applyIfNullary(eager, reducedResult, remainingArgs))));
                            } else {
                                // Not enough arguments available; back out
                                return pure(applyToArguments(original, args));
                            }
                        });
                    }
                });
            }

            @Override
            public Flow<Graph, Term> visit(Term.Variable instance) {
                // TODO: dereference variables
                return pure(applyToArguments(original, args));
            }
        });
    }

    private static  Term applyToArguments(Term function, LList<Term> args) {
        Term tcur = function;
        LList<Term> lcur = args;
        while (lcur != null) {
            tcur = Terms.apply(tcur, args.first);
            lcur = lcur.rest;
        }
        return tcur;
    }

    private static  boolean doRecurse(boolean eager, Term term) {
        boolean isLambda = term.accept(new Term.PartialVisitor<Boolean>() {
            @Override
            public Boolean otherwise(Term instance) {
                return false;
            }

            @Override
            public Boolean visit(Term.Function instance) {
                hydra.core.Function fun = instance.value;
                return fun.accept(new hydra.core.Function.PartialVisitor<Boolean>() {
                    @Override
                    public Boolean otherwise(hydra.core.Function instance) {
                        return false;
                    }

                    @Override
                    public Boolean visit(hydra.core.Function.Lambda instance) {
                        return true;
                    }
                });
            }
        });

        return eager && !isLambda;
    }

    public static  Flow<Graph, Term> reduce(boolean eager, Term term) {
        return reduce(eager, term, new HashMap<>(), null);
    }

    private static  Flow<Graph, Term> reduce(boolean eager,
                                                      Term original,
                                                      Map<Name, Term> env,
                                                      LList<Term> args) {
        return Rewriting.rewriteTermM(
                new Function<Function<Term, Flow<Graph, Term>>, Function<Term, Flow<Graph, Term>>>() {
                    @Override
                    public Function<Term, Flow<Graph, Term>> apply(
                            Function<Term, Flow<Graph, Term>> recurse) {
                        return new Function<Term, Flow<Graph, Term>>() {
                            @Override
                            public Flow<Graph, Term> apply(Term mid) {
                                // Do not recurse into lambda expressions,
                                // which will cause unexpected free variables to be encountered
                                Flow<Graph, Term> ready = doRecurse(eager, mid) ? recurse.apply(mid) : pure(mid);
                                return bind(ready, new Function<Term, Flow<Graph, Term>>() {
                                    @Override
                                    public Flow<Graph, Term> apply(Term inner) {
                                        return applyIfNullary(eager, inner, null);
                                    }
                                });
                            }
                        };
                    }
                }, Flows::pure, original);
    }

    private static  Flow<Graph, Term> reduceArg(boolean eager, Term arg) {
        // Reduce an argument only if evaluation is lazy (i.e. the argument may not already have been reduced)
        return eager ? pure(arg) : reduce(false, arg);
    }

    /**
     * Replace occurrences of a free variable (name) in a term with a given term.
     */
    public static  Term replaceFreeName(Name toReplace, Term replacement, Term body) {
        return Rewriting.rewriteTerm(recurse -> inner -> inner.accept(new Term.PartialVisitor<Term>() {
            @Override
            public Term otherwise(Term instance) {
                return recurse.apply(instance);
            }

            @Override
            public Term visit(Term.Function instance) {
                hydra.core.Function fun = instance.value;
                return fun.accept(new hydra.core.Function.PartialVisitor<Term>() {
                    @Override
                    public Term otherwise(hydra.core.Function instance) {
                        return recurse.apply(inner);
                    }

                    @Override
                    public Term visit(hydra.core.Function.Lambda instance) {
                        Lambda lam = instance.value;
                        return lam.parameter.equals(toReplace) ? inner : recurse.apply(inner);
                    }
                });
            }

            @Override
            public Term visit(Term.Variable instance) {
                return instance.value.equals(toReplace) ? replacement : instance;
            }
        }), a -> a, body);
    }
}